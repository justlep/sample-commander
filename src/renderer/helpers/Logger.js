let DEFAULT_LOG_LEVEL_NAME = 'WARN',
    CONSOLE_ENABLED = (typeof console !== 'undefined'),
    NOP = function(){},
    Logger,
    getConsoleMethodOrDefault = function(consoleMethodName) {
        /*eslint no-console:0 */
        return CONSOLE_ENABLED ? (console[consoleMethodName] || console['log'] || NOP) : NOP;
    },
    getLogLevelObjectByNameOrObject = function(nameOrObject) {
        let wishedLevelObject = (typeof nameOrObject === 'string') ? Logger.LEVEL[nameOrObject] : nameOrObject;
        for (let logLevelKey in Logger.LEVEL) {
            if (!Logger.LEVEL.hasOwnProperty(logLevelKey)) {
                continue;
            }
            let levelObject = Logger.LEVEL[logLevelKey];
            if (levelObject === wishedLevelObject) {
                return levelObject;
            }
        }
        Logger.LEVEL.ERROR.fn('Invalid log level - defaulting to ' + DEFAULT_LOG_LEVEL_NAME);
        return Logger.LEVEL[DEFAULT_LOG_LEVEL_NAME];
    };

Logger = {
    LEVEL: {
        DEBUG: {prio: 10, fn: getConsoleMethodOrDefault('debug')},
        INFO:  {prio: 20, fn: getConsoleMethodOrDefault('info')},
        DEV:   {prio: 25, fn: getConsoleMethodOrDefault('log')},
        WARN:  {prio: 30, fn: getConsoleMethodOrDefault('warn')},
        ERROR: {prio: 40, fn: getConsoleMethodOrDefault('error')},
        OFF:   {prio: 50, fn: NOP}
    },
    /** @static */
    debug: NOP,
    /** @static */
    info: NOP,
    /** @static */
    dev: NOP,
    /** @static */
    warn: NOP,
    /** @static */
    error: NOP,
    /**
     * Sets a new log level. If an invalid level is given, {@link DEFAULT_LOG_LEVEL_NAME} will be used.
     * @param {Object|string} newLogLevelNameOrObject - the name of the log level OR the LEVEL.* object itself
     * @static
     */
    setLogLevel: function(newLogLevelNameOrObject) {
        let newLogLevel = getLogLevelObjectByNameOrObject(newLogLevelNameOrObject),
            newPrio = newLogLevel.prio,
            logLevel,
            logMethodName,
            isMethodEnabled;
        for (let logLevelKey in Logger.LEVEL) {
            if (Logger.LEVEL.hasOwnProperty(logLevelKey)) {
                logLevel = Logger.LEVEL[logLevelKey];
                logMethodName = logLevelKey.toLowerCase();
                isMethodEnabled = (logLevel.prio >= newPrio);
                Logger[logMethodName] = (isMethodEnabled) ? logLevel.fn : NOP;
            }
        }
    }
};

Logger.setLogLevel(DEFAULT_LOG_LEVEL_NAME);

export default Logger;
