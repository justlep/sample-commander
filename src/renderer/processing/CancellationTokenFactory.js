
const NO_TOKEN = Symbol();

/**
 * A token that can be passed to asynchronous task queues to allow signalling them
 * that task processing should be stopped (cancelled) prematurely.
 * Only ONE token from the same {@link CancellationTokenFactory} is valid (in non-cancelled state) at a time.
 */
class CancellationToken {
    /**
     * @param {CancellationTokenFactory} factory
     */
    constructor(factory) {
        this._factory = factory;
    }

    /**
     * @returns {boolean}
     */
    isCancelled() {
        return !this._factory.isTokenActive(this);
    }
    
    cancel() {
        this._factory.cancelToken(this);
    }
}

/**
 * A Factory producing {@link CancellationToken} instances.
 * Only ONE token is valid (i.e. in non-cancelled state) at a time.
 * Generating a new token automatically cancels the previous one.
 */
export class CancellationTokenFactory {
    
    constructor() {
        this._activeToken = NO_TOKEN;
    }

    /**
     * Creates a new token, implicitly cancelling the previously active token.
     * @returns {CancellationToken}
     */
    newToken() {
        this._activeToken = new CancellationToken(this);
        return this._activeToken;
    }

    /**
     * Invalidates the currently active token. 
     */
    cancelCurrentToken() {
        this._activeToken = NO_TOKEN;
    }

    /**
     * Invalidates the given token if it is currently active.
     * @param {CancellationToken} token
     */
    cancelToken(token) {
        this.isTokenActive(token) && this.cancelCurrentToken();
    }
    
    /**
     * @param {*} token
     * @returns {boolean}
     */
    isTokenActive(token) {
        return token === this._activeToken;
    }
}
