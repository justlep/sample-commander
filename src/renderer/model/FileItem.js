import nodeCrypto from 'crypto'
import nodePath from 'path'
import { assertDirExists, stat } from '@/helpers/fileHelper'
import readdirp from 'readdirp'
import {NON_REACTIVE_PROPS, AUDIO_FILE_EXTENSIONS_REGEX} from '@/constants'
import Logger from '@/helpers/Logger'
import asserts from '@/helpers/asserts'

/**
 * Represents non-reactive (audio) file items.
 */
export default class FileItem {
    /**
     * @param {string} audioFilePathOrName - an audio file path or filename
     * @return {string}
     */
    static getSpectrogramPathByAudioFilePath(audioFilePathOrName) {
        return audioFilePathOrName + '.SPX.png';
    }
    
    /**
     * @param {FileItem[]} fileItems
     * @return {FileItem[]}
     */
    static sortList(fileItems) {
        return fileItems.sort((a, b) => a.path.localeCompare(b.path));
    }

    /**
     * @param {string} path
     * @param {boolean} [recursive=true]
     * @param {RegExp} [filenamePattern] - defaults to {@link AUDIO_FILE_EXTENSIONS_REGEX}
     * @param {number} [limit] - the maximum number of files to load (default = Number.MAX_SAFE_INTEGER}, 0 means no limit
     * @param {CancellationToken} [cancellationToken]
     * @param {FileItem[]} [onlySameSizesAs] - an optional array of file items whose sizes are used to pre-filter
     *                                         the scanned files. I.e. if given, only files contained in the given FileItem array
     *                                         will be returned. (Intended primarily for preparing duplicate checks)
     * @return {Promise<{dirItems: FileItem[], isLimitExceeded: Boolean}>}
     */
    static findInPath({path, filenamePattern = AUDIO_FILE_EXTENSIONS_REGEX, recursive = true, 
                          limit = Number.MAX_SAFE_INTEGER, cancellationToken, onlySameSizesAs}) {
        if (isNaN(limit)) {
            return Promise.reject('Invalid limit ' + limit);
        }
        const effectiveLimit = limit || Number.MAX_SAFE_INTEGER;
        const allowedFileSizesMap = Array.isArray(onlySameSizesAs) ? onlySameSizesAs.reduce((sizeMap, fileItem) => {
            sizeMap[fileItem.filesize] = 1;
            return sizeMap;
        }, {}) : null;
        
        return assertDirExists(path).then(() => new Promise((resolve, reject) => {
                let fileItems = [],
                    isLimitExceeded = false,
                    stream = readdirp(path, {
                        type: 'files',
                        fileFilter: allowedFileSizesMap ?
                            entryInfo => allowedFileSizesMap[entryInfo.stats.size || 0] && filenamePattern.test(entryInfo.basename) :
                            entryInfo => filenamePattern.test(entryInfo.basename),
                        depth: recursive ? Number.MAX_SAFE_INTEGER : 0,
                        alwaysStat: true,
                        /** (!) .asar files break readdirp -> see comment in {@link DirItem#loadRecursively} */ 
                        directoryFilter: ['!*.asar'] 
                    }),
                    finishedCallback = () => {
                        FileItem.sortList(fileItems);
                        resolve({fileItems, isLimitExceeded});
                    },
                    /**
                     * @param {Object} entryInfo - https://github.com/paulmillr/readdirp#entryinfo
                     */
                    onData = entryInfo => {
                        isLimitExceeded = fileItems.length >= effectiveLimit;
                        if (isLimitExceeded || (cancellationToken && cancellationToken.isCancelled())) {
                            Logger.debug('destroy file stream (cancelled or limit exceeded)');
                            stream.pause();
                            stream.destroy();
                            return;
                        }

                        let fileItem = new FileItem({
                            path: entryInfo.fullPath,
                            size: entryInfo.stats.size || 0,
                            mtime: entryInfo.stats.mtime.getTime(),
                            validate: !fileItems.length // validate first item to notice future breaking changes in readdirp
                        });

                        fileItems.push(fileItem);
                    };

            stream
                .on('data', onData)
                .on('warn', err => Logger.warn(err))
                .on('error', err => reject(err)) 
                //.on('end', finishedCallback)
                .on('close', () => {
                    Logger.debug('close FileItem stream');
                    finishedCallback();
                })
                .on('end', () => {
                    Logger.debug('end FileItem stream');
                    finishedCallback();
                });
            })
        );
    }

    /**
     * @param {string[]} paths
     * @returns {Promise<FileItem[]>}
     */
    static async getByPaths({paths}) {
        let fileItems = [];
        for (let filePath of paths) {
            try {
                let fileStat = await stat(filePath);
                if (fileStat.isFile()) {
                    fileItems.push(new FileItem({
                        path: filePath,
                        size: fileStat.size,
                        mtime: fileStat.mtime
                    }));
                }
            } catch (err) {
                Logger.dev('Ignoring missing file: %s', filePath);
            }
        }
        return fileItems;
    }
    
    /**
     * @param {string} path
     * @param {number} size
     * @param {number} mtime
     * @param {boolean} [validate] - set true to validate parameters and throw an error if invalid 
     */
    constructor({path, size = 0, mtime, validate}) {
        let resolvedPath = nodePath.resolve(path),
            idHash = nodeCrypto.createHash('sha1'),
            isAudioFile = AUDIO_FILE_EXTENSIONS_REGEX.test(resolvedPath),
            supportsSpectrograms = isAudioFile;

        if (validate) {
            asserts.assertNonEmptyString(path);
            asserts.assertNumberInRange(size, 0, Number.MAX_SAFE_INTEGER);
            asserts.assertNumberInRange(mtime, 0, Number.MAX_SAFE_INTEGER);
            // Logger.debug(path, size, mtime);
        }
        
        idHash.update(resolvedPath + '::' + size + '::' + mtime + '::');
        
        this[NON_REACTIVE_PROPS] = {
            id: idHash.digest('hex'),
            filename: nodePath.basename(resolvedPath),
            path: resolvedPath,
            parentDir: nodePath.join(resolvedPath, '..'),
            filesize: size,
            formattedMTime: new Date(mtime).toLocaleString(), // moment(mtime).format('DD.MM.YYYY - hh:mm:ss (ddd)'),
            formattedFilesize: (size / (1024 * 1024)).toFixed(1),
            isAudioFile,
            supportsSpectrograms,
            isDeleted: false,
            renamedIntoPath: null
        };
        
        this.metadata = {
            duration: '',
            sampleRate: ''
        };
        this.spectrogram = null;
        
        /** @type {String[]} */
        this._duplicatePaths = null;
    }
    
    get id() { return this[NON_REACTIVE_PROPS].id }
    get isAudioFile() { return this[NON_REACTIVE_PROPS].isAudioFile }
    get filename() { return this[NON_REACTIVE_PROPS].filename }
    get path() { return this[NON_REACTIVE_PROPS].path }
    get parentDir() { return this[NON_REACTIVE_PROPS].parentDir }
    get filesize() { return this[NON_REACTIVE_PROPS].filesize }
    get formattedMTime() { return this[NON_REACTIVE_PROPS].formattedMTime }
    get formattedFilesize() { return this[NON_REACTIVE_PROPS].formattedFilesize }
    get supportsSpectrograms() { return this[NON_REACTIVE_PROPS].supportsSpectrograms }
    
    get isDeleted() { return this[NON_REACTIVE_PROPS].isDeleted }
    get renamedIntoPath() { return this[NON_REACTIVE_PROPS].renamedIntoPath }

    hasMetadata() {
        return !!(this.metadata.duration || this.metadata.sampleRate);
    }

    /**
     * @returns {number|null} - the file's Nyquist frequency or null if the `metadata.sampleRate` isn't set 
     */
    getNyquistFrequency() {
        let sampleRate = this.metadata.sampleRate;
        return isNaN(sampleRate) ? null : sampleRate / 2;
    }
    
    /**
     * Returns all registered duplicate paths, but excluding its own path.
     * @return {String[]}
     */
    getDuplicatePathsWithoutSelf() {
        return (this._duplicatePaths || []).filter(path => path !== this.path);
    }
    
    /**
     * @param {FileItem} fileItem
     */
    addDuplicate(fileItem) {
        if (this._duplicatePaths) {
            this._duplicatePaths.push(fileItem.path);
        } else {
            this._duplicatePaths = [fileItem.path];
        }
    }
    
    resetDuplicates() {
        this._duplicatePaths = null;
    }
    
    /** @type {boolean} */
    hasDuplicates() {
        return !!(this._duplicatePaths && this._duplicatePaths.length);
    }
    
    /** @type {boolean} */
    isUnique() {
        return !this._duplicatePaths || !this._duplicatePaths.length;
    }

    /**
     * @return {string}
     */
    getSpectrogramFilename() {
        return FileItem.getSpectrogramPathByAudioFilePath(this.filename);
    }
    
    /**
     * @return {string}
     */
    getSpectrogramPath() {
        return FileItem.getSpectrogramPathByAudioFilePath(this.path);
    }

    setDeleted() {
        this[NON_REACTIVE_PROPS].isDeleted = true;
    }

    /**
     * Marks this file item to be renamed, so it can be cleaned up and re-looked-up.
     * @param {string} newPath
     */
    setRenamed(newPath) {
        this[NON_REACTIVE_PROPS].renamedIntoPath = newPath;
    }
    
    /**
     * @override
     */
    toString() {
        return 'FileItem['+ this.id +']';
    }
}


/**
 * @typedef {Object} Readdirp_EntryInfo
 * @property {string} parentDir - directory in which entry was found (relative to given root)
 * @property {string} fullParentDir - full path to parent directory
 * @property {string} name - name of the file/directory
 * @property {string} path - path to the file/directory (relative to given root)
 * @property {string} fullPath - full path to the file/directory found
 * @property {Object} stat - built in stat object
 */
