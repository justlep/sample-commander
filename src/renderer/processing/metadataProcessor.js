import {execFile} from 'child_process';
import Logger from '@/helpers/Logger'
import {format} from 'util';

/**
 * Determines metadata (e.g. sample rate) for this file using the ffprobe executable.
 * If the metadata is in the cache, that cached version is returned.
 *
 * @param {FileItem} fileItem
 * @param {string} ffprobeExecutablePath
 * @returns {Promise<{duration: String, sampleRate: Number}>}
 */
export async function getMetadataForFileItem(fileItem, ffprobeExecutablePath) {
    if (!ffprobeExecutablePath) {
        throw new Error('Invalid ffprobeExecutablePath');
    }

    let metadataJson;
    
    try {
        let metadataJsonString = await new Promise((resolve, reject) => {
                Logger.dev('Running ffprobe for metadata of %s', fileItem.path);
    
                execFile(ffprobeExecutablePath, [
                    '-hide_banner',
                    '-show_streams',
                    '-select_streams', 'a:0',
                    // '-sexagesimal',  // <-- for duration getting formatted like "0:09:39.396984" 
                    '-show_entries', 'stream=duration,duration_ts,sample_rate,time_base',
                    '-print_format', 'json',
                    fileItem.path
                ], {
                    timeout: 3000,
                    shell: false,
                    encoding: 'utf8',
                    stdio: ['pipe', 'pipe', 'ignore']
                }, (err, stdout, stderr) => {
                    if (err) {
                        reject(format('ffprobe failed to load metadata for %s, error: %s', fileItem.path, stderr));
                    } else {
                        resolve('' + stdout);
                    }
                });
            }),
            json = JSON.parse(metadataJsonString),
            firstStream = json && json.streams[0],
            sampleRate,
            duration,
            durationInMins;
        
        if (firstStream) {
            sampleRate = firstStream && parseInt(firstStream['sample_rate'], 10);
            
            let durationInSecs = parseFloat(firstStream.duration);

            // Logger.dev(firstStream);
            
            if (!isNaN(durationInSecs)) {
                // duration is in minutes
                let hours = (durationInSecs / 3600) << 0, 
                    mins = ((durationInSecs - (hours * 3600)) / 60) << 0, 
                    secs = (durationInSecs % 60) << 0,
                    parts = [];

                if (durationInSecs >= 1) {
                    if (hours) {
                        parts.push(hours + 'h');
                    }
                    if (mins) {
                        parts.push(mins + 'm');
                    }
                    if (secs) {
                        parts.push(secs + 's');
                    }
                } else {
                    parts.push(((durationInSecs * 1000) << 0) + 'ms');
                }
                duration = parts.join(' ');
                durationInMins = durationInSecs / 60;
            }
        }

        metadataJson = {
            sampleRate: sampleRate || null, 
            duration,
            durationInMins
        };
        
    } catch (err) {
        Logger.error('Failed to process metadata: %o', err);
        throw new Error(format('Failed to process metadata JSON  for %s, error: %s', fileItem.path, err));
    }
    
    // Logger.dev('Finished metadata for %s -> %o', fileItem.filename, metadataJson);
    return metadataJson;
}
