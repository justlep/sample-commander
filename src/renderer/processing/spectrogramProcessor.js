import {assertFileExists, pathExists} from '@/helpers/fileHelper'
import {execFile} from 'child_process'
import {format} from 'util'
import Logger from '@/helpers/Logger'
import {convertToMelScale} from '@/processing/spectrogramScaler'
import nodePath from 'path'

const FFMPEG_TIMEOUT_IN_MILLIS = 60 * 1000;
const FINAL_SPECTROGRAM_FREQUENCY_RANGE = 20000;
const FINAL_SPECTROGRAM_WIDTH = 800;
const FINAL_SPECTROGRAM_HEIGHT = 256;
const LINEAR_SPECTROGRAM_MIN_VERTICAL_SOURCE_PIXELS_CONTAINING_FINAL_FREQ_RANGE = 1024;
const KEEP_LINEAR_SPECTROGRAM_IMAGE = false;

// keep this local for now, maybe put into config store later 
const SPECTROGRAM_BRIGTHNESS_FACTOR = 1.25;
const SPECTROGRAM_SATURATION_FACTOR = 1.25;

/**
 * @param {FileItem} fileItem
 * @returns {number} - The default frequency range of the spectrogram produced by ffmpeg showspectrumpic
 *                     (when used without problematic (slow) start/stop bandpass parameters).
 * @private
 */
function _getLinearSpectrogramFrequencyRange(fileItem) {
    let freqRange = fileItem.getNyquistFrequency();
    if (!freqRange) {
        throw new Error('FileItem ' + fileItem + ' has invalid Nyquist frequency ' + freqRange);
    }
    return freqRange;
}

/**
 * Generates a linear spectrogram image for the given audio file's path using the given ffmpeg executable.
 * If a linear spectrogram already exists, that one is returned.
 *
 * (!) Had some troubles with ffmpeg showspectrumpic here:
 *
 *    (1) Using bandpass parameters "start=0:stop=20000" causes processing slow-down of factor > 7
 *        -> Solution = don't use start/stop bandpass, instead generate a larger spectrogram,
 *        then use only the "interesting" spectrum parts during post-processing (Mel-scaling)
 *    (2) ffmpeg 4.1 showspectrumpic produces faulty spectrograms for different heights,
 *       -> Solution = use only heights 1024, 2048, 4096 which seemed to work fine
 *    (3) ffmpeg 4.1 (and other 2018 builds) crashes 
 *        when trying showspectrumpic height > 1024 for files of 1GB+ (height 2048 at 999MB works)
 *
 *    Documentation for ffmpeg showspectrumpic:
 *    --> https://ffmpeg.org/ffmpeg-filters.html#toc-showspectrumpic
 *
 * @param {FileItem} fileItem
 * @param {string} ffmpegExecutablePath
 * @param {string} [linearSpectrogramDir] - path to a folder for temporary files 
 *                                          If given, intermediate (linear) spectrogram images will be saved here,
 *                                          otherwise in the fileItems's parentDir
 * @return {Promise<string>} - resolve with the existing or newly generated linear spectrogram path, rejects if failed
 * @private
 */
function _getOrGenerateLinearSpectrogram(fileItem, ffmpegExecutablePath, linearSpectrogramDir) {
    const linearSpectrogramFilename = 'SPX_' + fileItem.id + '.png';
    const linearSpectrogramPath = nodePath.join(linearSpectrogramDir || fileItem.parentDir, linearSpectrogramFilename);

    if (!ffmpegExecutablePath) {
        return Promise.reject('Invalid ffmpegExecutable ');
    }
    
    return assertFileExists(linearSpectrogramPath)
        .then(() => {
            Logger.dev('Using existing linear spectrogram for %s', fileItem);
            return linearSpectrogramPath;
        })
        .catch(() => new Promise(async (resolve, reject) => {
            let linearSpectrogramFrequencyRange;
            
            try {
                linearSpectrogramFrequencyRange = _getLinearSpectrogramFrequencyRange(fileItem);
            } catch (err) {
                reject(err);
                return;
            }
            
            let minLinearSpectrogramHeight = (linearSpectrogramFrequencyRange / FINAL_SPECTROGRAM_FREQUENCY_RANGE) *
                    LINEAR_SPECTROGRAM_MIN_VERTICAL_SOURCE_PIXELS_CONTAINING_FINAL_FREQ_RANGE,
                ffmpegSafeHeightFactor = Math.min(4, Math.ceil(minLinearSpectrogramHeight / 1024)),
                linearSpectrogramHeight = 1024 * (ffmpegSafeHeightFactor === 3 ? 4 : ffmpegSafeHeightFactor);

            if (fileItem.filesize > 1024 * 1024 * 1023) {
                // for some strange reason, ffmpeg spectrumpic height > 1024 crashes for files of 1GB+ 
                linearSpectrogramHeight = 1024;
            }
            
            Logger.dev('Running ffmpeg for linear spectrogram of %s', fileItem.path);
            
            execFile(ffmpegExecutablePath, [
                '-hide_banner',
                '-y',
                '-i',
                fileItem.filename,
                '-lavfi',
                `showspectrumpic=size=${FINAL_SPECTROGRAM_WIDTH}x${linearSpectrogramHeight}:color=fire:scale=log:legend=0`,
                linearSpectrogramPath
            ], {
                cwd: fileItem.parentDir,
                timeout: FFMPEG_TIMEOUT_IN_MILLIS
            }, (err, stdout, stderr) => {
                if (err) {
                    reject( format('Failed to generate %s. Error is:\n%s', linearSpectrogramPath, stderr) );
                } else {
                    Logger.dev('Created linear spectrogram: %s', linearSpectrogramPath);
                    resolve(linearSpectrogramPath);
                }
            });
        }));
}

/**
 * Create a spectrogram image (in Mel scale) for the given audio file path. 
 * If non exists, a new spectrogram image will be generated.
 *
 * @param {FileItem} fileItem
 * @param {string} ffmpegExecutablePath - the path to the ffmpeg.exe (windows) or respective executable on other OS
 * @param {string} [linearSpectrogramDir] - if given, intermediate (linear) spectrogram images will be saved here,
 *                                          otherwise in the fileItems's parentDir          
 * @return {Promise<string>} - the path to the generated or already existing spectrogram image
 */
export async function getSpectrogramForFileItem(fileItem, ffmpegExecutablePath, linearSpectrogramDir) {
    const spectrogramPath = fileItem.getSpectrogramPath();

    let validSpectrogramPath = null;

    if (await pathExists(spectrogramPath)) {
        validSpectrogramPath = spectrogramPath;
    } else {
        try {
            const linearSpectrogramPath = await _getOrGenerateLinearSpectrogram(fileItem, ffmpegExecutablePath, linearSpectrogramDir);

            validSpectrogramPath = await new Promise((resolve, reject) => {
                convertToMelScale({
                    linearSpectrogramPath,
                    targetSpectrogramPath: spectrogramPath,
                    targetHeight: FINAL_SPECTROGRAM_HEIGHT,
                    deleteLinearSpectrogramAfterSuccess: !KEEP_LINEAR_SPECTROGRAM_IMAGE,
                    linearSpectrogramFrequencyRange: _getLinearSpectrogramFrequencyRange(fileItem),
                    targetFrequencyRange: FINAL_SPECTROGRAM_FREQUENCY_RANGE,
                    brightnessFactor: SPECTROGRAM_BRIGTHNESS_FACTOR,
                    saturationFactor: SPECTROGRAM_SATURATION_FACTOR
                })
                    .then(successfullyScaledSpectroPath => {
                        Logger.dev('Converted spectrogram to Mel scale');
                        resolve(successfullyScaledSpectroPath);
                    }).catch(err => {
                    Logger.error(err);
                    reject('Failed to convert spectrogram to Mel scale');
                });
            });

        } catch (err) {
            throw err;
        }
    }

    return validSpectrogramPath;
}
