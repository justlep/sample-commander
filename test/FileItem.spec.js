import Logger from '@/helpers/Logger'
import FileItem from '@/model/FileItem'
import {CancellationTokenFactory} from '@/processing/CancellationTokenFactory'
import path from 'path'

const TEST_FILES_DIR_PATH = path.join(__dirname, 'test-files');

/**
 * @param {FileItem} fileItem
 */
function fileItemToJson(fileItem) {
    let {filename, filesize, isAudioFile, supportsSpectrograms} = fileItem;
    return {filename, filesize, isAudioFile, supportsSpectrograms};
}

describe('FileItem', () => {

    beforeAll(() => {
        Logger.setLogLevel(Logger.LEVEL.OFF)
    });
    
    it('can scan for audio files recursively (by default)', async () => {
        let {fileItems, isLimitExceeded} = await FileItem.findInPath({path: TEST_FILES_DIR_PATH});

        expect(isLimitExceeded).toBe(false);
        expect(Array.isArray(fileItems)).toBe(true);

        let nonAudioFileItem = fileItems.find(it => !it.isAudioFile); 
        expect(nonAudioFileItem).toBeUndefined();
        
        expect(fileItems.find(it => it.filename.includes('-in-subdir.'))).toBeInstanceOf(FileItem);
        
        let itemsJson = fileItems.map(fileItemToJson);
        expect(itemsJson).toMatchSnapshot();
    });


    it('can scan for audio files non-recursively', async () => {
        let {fileItems, isLimitExceeded} = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: false});

        expect(isLimitExceeded).toBe(false);
        expect(Array.isArray(fileItems)).toBe(true);
        expect(fileItems.length).toBeGreaterThan(0);

        let nonAudioFileItem = fileItems.find(it => !it.isAudioFile);
        expect(nonAudioFileItem).toBeUndefined();

        expect(fileItems.find(it => it.filename.includes('-in-subdir.'))).toBeUndefined();
        
        let itemsJson = fileItems.map(fileItemToJson);
        expect(itemsJson).toMatchSnapshot();
    });
    
    it('can scan for files with a limit', async () => {
        let {fileItems, isLimitExceeded} = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true});
        
        const LIMIT = 5;
        const totalFilesUnlimited = fileItems.length; 
        expect(totalFilesUnlimited).toBeGreaterThan(LIMIT);
        expect(isLimitExceeded).toBe(false);

        // with limit below actually present files
        let retObj = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true, limit: LIMIT});
        expect(Array.isArray(retObj.fileItems)).toBe(true);
        expect(retObj.fileItems.length).toBe(LIMIT);
        expect(retObj.isLimitExceeded).toBe(true);

        // with limit that matches exactly the # of present files (i.e. limit reached, but not exceeded)
        
        retObj = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true, limit: totalFilesUnlimited});
        expect(Array.isArray(retObj.fileItems)).toBe(true);
        expect(retObj.fileItems.length).toBe(totalFilesUnlimited);
        expect(retObj.isLimitExceeded).toBe(false);
    });
    
    it('can scan for files with a custom filenamePattern (regex)', async () => {
        let retObj = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true, filenamePattern: /\.txt$/});
        expect(Array.isArray(retObj.fileItems)).toBe(true);
        expect(retObj.fileItems.map(fileItemToJson)).toMatchSnapshot();
    });
    
    it ('can be interrupted with scanning by a given CancellationToken', async () => {
        const tokenFactory = new CancellationTokenFactory();
        let cancellationToken = tokenFactory.newToken(),
            retObj = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true, cancellationToken});

        expect(cancellationToken.isCancelled()).toBe(false);
        expect(retObj.fileItems.length).toBeGreaterThan(0);
        expect(cancellationToken.isCancelled()).toBe(false);
        
        let token2 = tokenFactory.newToken();
        expect(cancellationToken.isCancelled()).toBe(true); // old token is invalidated by generating token2
        expect(token2.isCancelled()).toBe(false);
        token2.cancel();
        expect(token2.isCancelled()).toBe(true);
        retObj = await FileItem.findInPath({path: TEST_FILES_DIR_PATH, recursive: true, cancellationToken: token2});
        expect(retObj.fileItems.length).toBe(0);
    });
});
