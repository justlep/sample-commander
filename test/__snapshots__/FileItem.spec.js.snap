// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileItem can scan for audio files non-recursively 1`] = `
Array [
  Object {
    "filename": "20k-sweep-11k.wav",
    "filesize": 16584,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-192k.wav",
    "filesize": 288072,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-22k.wav",
    "filesize": 33122,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-44k.wav",
    "filesize": 66200,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k-copy.wav",
    "filesize": 72050,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k.wav",
    "filesize": 72050,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-96k.wav",
    "filesize": 144056,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
]
`;

exports[`FileItem can scan for audio files recursively (by default) 1`] = `
Array [
  Object {
    "filename": "20k-sweep-11k.wav",
    "filesize": 16584,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-192k.wav",
    "filesize": 288072,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-22k.wav",
    "filesize": 33122,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-44k.wav",
    "filesize": 66200,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k-copy.wav",
    "filesize": 72050,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k.wav",
    "filesize": 72050,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-96k.wav",
    "filesize": 144056,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k-in-subdir.mp3",
    "filesize": 4461,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
  Object {
    "filename": "20k-sweep-48k-in-subdir.wav",
    "filesize": 72050,
    "isAudioFile": true,
    "supportsSpectrograms": true,
  },
]
`;

exports[`FileItem can scan for files with a custom filenamePattern (regex) 1`] = `
Array [
  Object {
    "filename": "small-file.txt",
    "filesize": 446,
    "isAudioFile": false,
    "supportsSpectrograms": false,
  },
]
`;
